Frontend changes for logged-out experience (User checkpoints project)

IamEzio opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR covers the frontend part for logged-out experience (User Checkpoints project) ([TDD](https://docs.google.com/document/d/1eWHs46cOHcm7NuQl8pM9Dada9nTgg3GoCQLgQJSNmIY/edit#heading=h.liiqbmwmz8j8))
2. This PR does the following: 
    - Adds user checkpoints feature flag.
    - Adds functionality for a logged-out learner to save and resume exploration using a unique progress url.
    - Allows logged-out learner to log in and resume their exploration.

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

#### Checkpoints feature flag

https://user-images.githubusercontent.com/89375125/177056509-e9c705b6-631b-4b20-9146-20acac0fd11b.mp4

#### Logged-out learner uses progress URL

https://user-images.githubusercontent.com/89375125/177108791-3d2a8ce8-cd04-4283-aa4a-2366ea3d82e2.mp4


#### Logged-out learner signs in

https://user-images.githubusercontent.com/89375125/177056550-3912efbf-3e69-4c2f-a46a-161f17f49b1d.mp4

#### Logged-out learner uses invalid progress URL

https://user-images.githubusercontent.com/89375125/177157829-b4b2e67b-a3d6-4364-9e8d-f4d9e9ee82e5.mp4

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

https://user-images.githubusercontent.com/89375125/177056561-510fb149-05c9-4f4d-a01b-1f394c8ffb77.mp4


<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

https://user-images.githubusercontent.com/89375125/177056565-53b3ffa0-15e7-4b91-8d5e-c4512fa0532d.mp4


<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

https://user-images.githubusercontent.com/89375125/177056581-ae6b194b-ca37-4d2e-bd5d-b349e28a03aa.mp4


<!--
If the PR changes the UI in any of the files listed in rtl_css.py (i.e, those that have
a separate .rtl.css file for styling), make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

IamEzio committed:
added utility methods and their tests

IamEzio committed:
Merge branch 'oppia:develop' into user-checkpoint-front

IamEzio committed:
added methods for handling localstorage

IamEzio committed:
added remaining utility methods and updated progress save workflow

IamEzio committed:
added tests

IamEzio committed:
nits

IamEzio committed:
added basic modal

IamEzio committed:
Merge branch 'oppia:develop' into user-checkpoint-front

IamEzio committed:
Merge branch 'user-checkpoint-front' of https://github.com/IamEzio/oppia into user-checkpoint-front

IamEzio committed:
added uid-box

IamEzio committed:
Merge branch 'oppia:develop' into user-checkpoint-front

IamEzio committed:
Merge branch 'oppia:develop' into user-checkpoint-front

IamEzio committed:
Merge branch 'oppia:develop' into user-checkpoint-front

IamEzio committed:
Merge branch 'user-checkpoint-front' of https://github.com/IamEzio/oppia into user-checkpoint-front

IamEzio committed:
completed root functionality

IamEzio committed:
completed logged-out sync functionality

IamEzio committed:
updated progress url

IamEzio committed:
added checkpoint feature flag

IamEzio committed:
css fix

IamEzio committed:
functionality completed

IamEzio committed:
old tests fix

IamEzio committed:
added tests

IamEzio committed:
added more tests

IamEzio committed:
nits

IamEzio committed:
lint fix

IamEzio committed:
css lint fix

IamEzio committed:
css lint fix

IamEzio committed:
nit

IamEzio commented:
@paprajapati9 @seanlip PTAL! Thanks

IamEzio committed:
rtl css command done

paprajapati9 commented:
@IamEzio Featurewise things look really good to me, so kudos for that! I just have 1 major issue, that is the save progress modal. I think it's a bit off, can you consider resducing it's size, as it looks odd having a lot of empty space in it. 
Did you refer to this [mock](https://www.figma.com/file/NRJaGClL9EUJBAhugi4oPQ/Oppia-User-Checkpoints?node-id=814%3A199)  for the modal ?

IamEzio committed:
Merge branch 'oppia:develop' into user-checkpoint-front

IamEzio committed:
added backend test

IamEzio committed:
test fix

IamEzio commented:
> @IamEzio Featurewise things look really good to me, so kudos for that! I just have 1 major issue, that is the save progress modal. I think it's a bit off, can you consider resducing it's size, as it looks odd having a lot of empty space in it. Did you refer to this [mock](https://www.figma.com/file/NRJaGClL9EUJBAhugi4oPQ/Oppia-User-Checkpoints?node-id=814%3A199) for the modal ?

Hi @paprajapati9! The space is for the email and text functionalities that might be added in the future. If it is needed, a separate PR can be raised for the modal as we discussed. Thanks!

IamEzio committed:
css update

IamEzio commented:
> @IamEzio Featurewise things look really good to me, so kudos for that! I just have 1 major issue, that is the save progress modal. I think it's a bit off, can you consider resducing it's size, as it looks odd having a lot of empty space in it. Did you refer to this [mock](https://www.figma.com/file/NRJaGClL9EUJBAhugi4oPQ/Oppia-User-Checkpoints?node-id=814%3A199) for the modal ?

I have updated it now. Please refer to the video under the heading **Logged-out learner uses progress url** in the **Proof that changes are correct** section. Thanks!

paprajapati9 commented on core/templates/app.constants.ts:
```diff
@@ -50,6 +50,8 @@ export const AppConstants = {
   EXPLORATION_DATA_URL_TEMPLATE: '/explorehandler/init/<exploration_id>',
   EXPLORATION_VERSION_DATA_URL_TEMPLATE: (
     '/explorehandler/init/<exploration_id>?v=<version>'),
+  EXPLORATION_PROGRESS_UID_URL_TEMPLATE: (
```
Are you calling this uid or pid? Use a single name everywhere.

IamEzio replied:
Fixed

paprajapati9 commented on core/templates/pages/exploration-player-page/templates/lesson-information-card-modal.component.html:
```diff
@@ -48,13 +48,22 @@ <h4 class="oppia-exploration-description" *ngIf="isHackyExpDescTranslationDispla
+                    *ngIf="!userIsLoggedIn"
+                    [ngClass]="{'disabled oppia-disabled-link': !completedWidth}"
+                    (click)="saveLoggedOutProgress()">
+              Save Progress
```
Please make this I18N compatible.

IamEzio replied:
Done

paprajapati9 commented on core/templates/pages/exploration-player-page/templates/lesson-information-card-modal.component.html:
```diff
@@ -103,13 +112,22 @@ <h4 class="oppia-exploration-description" *ngIf="isHackyExpDescTranslationDispla
+                    *ngIf="!userIsLoggedIn"
+                    [ngClass]="{'disabled oppia-disabled-link': !completedWidth}"
+                    (click)="saveLoggedOutProgress()">
+              Save Progress
```
Ditto

IamEzio replied:
Done

paprajapati9 commented on core/templates/pages/exploration-player-page/templates/lesson-information-card-modal.component.html:
```diff
@@ -125,6 +143,36 @@ <h4 class="oppia-exploration-description" *ngIf="isHackyExpDescTranslationDispla
+     *ngIf="saveProgressMenuIsShown && !userIsLoggedIn">
+  <h3>Save Progress</h3>
+  <div class="sign-in-box">
+    <p>Your learning progress will be automatically saved if you have an account.</p>
```
Please make all the strings in this modal I18N compatible.

IamEzio replied:
Done!

paprajapati9 reviewed (commented):
Thanks @IamEzio! Just few minor comments from my side. Also I do not see a proof video in case the progress id given is invalid. Can you please attach that as well.

paprajapati9 commented:
@kevintab95 @srijanreddy98 @gp201 Can you guys please review this asap! We need to get this PR merged for the July release so that the whole checkpoints feature can go in this release. Thanks!

kevintab95 commented on core/controllers/reader.py:
```diff
@@ -1794,3 +1794,21 @@ def post(self, exploration_id):
+
+    @acl_decorators.open_access
+    def get(self):
+        self.render_json({
+            'checkpoints_feature_is_enabled': (
+                config_domain.CHECKPOINTS_FEATURE_IS_ENABLED.value)
+        })
```
Instead of using config domain to enable/disable this feature, can we use the dedicated "feature gating" service in the codebase (See `platform_feature_services.py`)?

seanlip replied:
I think the plan is actually to use config_domain and the admin page for this, since it's more well-tested. Let's start with that for now since this needs to go out soonish.

kevintab95 replied:
Sounds good, if we plan to do it at a later time, please add a TODO comment + issue. Thanks!

IamEzio replied:
Hi Kevin! I have made the issue #15682 regarding the same. Also added a TODO in config_domain.

kevintab95 commented on core/templates/domain/exploration/editable-exploration-backend-api.service.spec.ts:
```diff
@@ -317,6 +318,95 @@ describe('Editable exploration backend API service', function() {
       { status: 200, statusText: 'Success.'});
     flushMicrotasks();
 
+    editableExplorationBackendApiService.
```
Can we have a separate test for the logged out experience?

IamEzio replied:
Done!

kevintab95 commented on core/templates/domain/exploration/editable-exploration-backend-api.service.ts:
```diff
@@ -128,14 +128,56 @@ export class EditableExplorationBackendApiService {
+
+  async changeLoggedOutProgressToLoggedInProgressAsync(
+      explorationId: string,
+      pid: string
```
It is a bit unclear what pid is..

IamEzio replied:
Changed it to `uniqueProgressUrlId` now.

kevintab95 commented on core/templates/domain/exploration/read-only-exploration-backend-api.service.spec.ts:
```diff
@@ -134,6 +134,24 @@ describe('Read only exploration backend API service', () => {
   }));
 
+  it('should successfully fetch an existing exploration from a unique' +
+    ' url progress id', fakeAsync(() => {
```
```suggestion
    ' URL progress id', fakeAsync(() => {
```

IamEzio replied:
Done

kevintab95 commented on core/templates/domain/exploration/read-only-exploration-backend-api.service.ts:
```diff
@@ -67,14 +72,33 @@ export class ReadOnlyExplorationBackendApiService {
 
   private async _fetchExplorationAsync(
-      explorationId: string, version: number | null
+      explorationId: string, version: number | null, pid: string | null = null
```
Ditto: Please use a more meaningful variable name. It is unclear what pid means.

kevintab95 replied:
Ditto elsewhere regarding pid.

IamEzio replied:
Done

kevintab95 commented on core/templates/pages/exploration-player-page/services/exploration-player-state.service.spec.ts:
```diff
@@ -655,4 +655,24 @@ describe('Exploration Player State Service', () => {
+        unique_progress_url_id: '123456'
+      }));
+    expect(explorationPlayerStateService.getUniqueProgressUrlId()).toBeNull();
+    explorationPlayerStateService.setLastCompletedCheckpoint('abc');
+    explorationPlayerStateService.setUniqueProgressUrlId();
```
Should we have expect statements to check if the URL ID was set?

IamEzio replied:
Yes, I missed it. Added it now. Thanks!

kevintab95 commented on core/templates/pages/exploration-player-page/services/exploration-player-state.service.spec.ts:
```diff
@@ -655,4 +655,24 @@ describe('Exploration Player State Service', () => {
+      .toBeTrue();
+  });
+
+  it('should set unique progress url id correctly', () => {
```
```suggestion
  it('should set unique progress URL id correctly', () => {
```

IamEzio replied:
Done!

kevintab95 commented on core/templates/pages/exploration-player-page/templates/lesson-information-card-modal.component.css:
```diff
@@ -135,13 +135,27 @@
 
+.oppia-lesson-info-card .save-progress-btn {
+  background-color: #00645c;
+  color: white;
```
Here and below: Please use either HEX or RGB values (see [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide#css).)

IamEzio replied:
Done!

kevintab95 commented on core/templates/pages/exploration-player-page/templates/lesson-information-card-modal.component.html:
```diff
@@ -125,6 +143,36 @@ <h4 class="oppia-exploration-description" *ngIf="isHackyExpDescTranslationDispla
+    <p class="guide-text">Use the link below to save progress for 72 hours.<br>You may lose your progress after 72 hours.</p>
+    <p class="copy-text">Write or copy the link below</p>
+    <div class="uid-copy" [ngStyle]="{ 'flex-direction': isLanguageRTL() ? 'row-reverse' : 'row' }">
+      <div id="unique-progress-id">{{ loggedOutProgressUniqueUrl }}</div>
```
Why do we need to use the ID attribute here?

IamEzio replied:
Updated it now.

kevintab95 commented on core/templates/pages/exploration-player-page/templates/lesson-information-card-modal.component.spec.ts:
```diff
@@ -228,6 +235,25 @@ describe('Lesson Information card modal component', () => {
   }));
 
+  it('should correctly set logged-out progress learner url ' +
+    'when unique progress url id exists', fakeAsync (() => {
```
```suggestion
    'when unique progress URL ID exists', fakeAsync (() => {
```

IamEzio replied:
Done!

kevintab95 commented on core/templates/pages/exploration-player-page/templates/lesson-information-card-modal.component.spec.ts:
```diff
@@ -253,4 +279,46 @@ describe('Lesson Information card modal component', () => {
+      loggedOutProgressUrl);
+  });
+
+  it('should store unique progress url id when login button is clicked',
```
```suggestion
  it('should store unique progress URL ID when login button is clicked',
```

IamEzio replied:
Done

kevintab95 commented on core/templates/pages/exploration-player-page/templates/lesson-information-card-modal.component.spec.ts:
```diff
@@ -253,4 +279,46 @@ describe('Lesson Information card modal component', () => {
+    expect(componentInstance.loggedOutProgressUniqueUrlId).toEqual('abcdef');
+  }));
+
+  it('should correctly copy progress url', () => {
```
```suggestion
  it('should correctly copy progress URL', () => {
```

IamEzio replied:
Done

kevintab95 commented on core/templates/pages/exploration-player-page/templates/lesson-information-card-modal.component.ts:
```diff
@@ -166,4 +179,36 @@ export class LessonInformationCardModalComponent extends ConfirmOrCancelModal {
+            '/progress/' + this.loggedOutProgressUniqueUrlId);
+        });
+    }
+    this.saveProgressMenuIsShown = true;
```
Does this depend on the successful completion of the call above?

IamEzio replied:
No, it does not. It is simply there to show the save progress menu.

kevintab95 replied:
Should the menu be shown / hidden when the call above fails?

IamEzio replied:
No. The call is meant to be executed only once to fetch the logged out progress URL ID from the backend. For the rest of the cases, this function would only toggle the save progress menu. Regarding the failure of the above call due to any internal server error, we would still want the user to see the save progress menu which contains a prompt that his progress will be saved automatically if he signs in.

kevintab95 commented on core/templates/pages/exploration-player-page/templates/lesson-information-card-modal.component.ts:
```diff
@@ -166,4 +179,36 @@ export class LessonInformationCardModalComponent extends ConfirmOrCancelModal {
+  }
+
+  copyProgressUrl(): void {
+    navigator.clipboard.writeText(this.loggedOutProgressUniqueUrl);
```
Can we use the Angular CDK clipboard instead? We seem to use it already in the `beam-job-output-dialog` component.

IamEzio replied:
Done!

kevintab95 commented on core/templates/services/contextual/url.service.spec.ts:
```diff
@@ -356,6 +356,17 @@ describe('Url Service', () => {
     expect(urlService.getExplorationVersionFromUrl()).toBe(null);
   });
 
+  it('should correctly retrieve unique progress id from the url', () => {
```
```suggestion
  it('should correctly retrieve unique progress ID from the URL', () => {
```

IamEzio replied:
Done!

kevintab95 commented on core/templates/services/local-storage.service.spec.ts:
```diff
@@ -204,5 +204,32 @@ describe('LocalStorageService', () => {
       expect(callbackFnSpy).toHaveBeenCalled();
     });
+
+    it('should correctly save unique progress url id', () => {
```
Ditto here and below: url -> URL, id -> ID

IamEzio replied:
Done!

kevintab95 reviewed (changes requested):
Thanks @IamEzio, nice work! Took a review pass and left some comments PTAL

IamEzio committed:
added i18 compatibility

IamEzio commented:
@paprajapati9 PTAL. I have added the video to show when a learner uses an invalid URL. Also added i18 compatibility to the modal and other changes you requested.

paprajapati9 commented on assets/i18n/qqq.json:
```diff
@@ -846,7 +847,13 @@
 	"I18N_RESTART_EXPLORATION_BUTTON": "Text which appears on the Restart button in the lesson info modal of the learner checkpoints progress.",
 	"I18N_REVIEW_TEST_PAGE_BREADCRUMB": "Text displayed in the review test page. - Text shown in the top left corner of the nav bar.",
 	"I18N_REVIEW_TEST_PAGE_TITLE": "Text displayed on the browser tab when on the review test page.",
+    "I18N_SAVE_EXPLORATION_PROGRESS_TEXT_1": "Text displayed on the save progress menu of exploration lesson info modal telling user that his progress will be automatically saved if he signs in.",
```
Indentation is off.

IamEzio replied:
Fixed

paprajapati9 reviewed (approved):
Thanks @IamEzio! LGTM. Just one comment.

IamEzio committed:
renamed paramters and updated tests

IamEzio commented:
@kevintab95 PTAL! Thanks

vojtechjelinek reviewed (approved):
LGTM!

IamEzio committed:
added TODO

IamEzio committed:
fixed test

IamEzio commented:
@kevintab95 PTAL! Thanks

nithusha21 commented on assets/i18n/qqq.json:
```diff
@@ -846,7 +847,13 @@
+	"I18N_SAVE_EXPLORATION_PROGRESS_TEXT_2": "Text displayed on the save progress menu of exploration lesson info modal asking the user whether he already has an account.",
+	"I18N_SAVE_EXPLORATION_PROGRESS_TEXT_3": "Text displayed on the save progress menu of exploration lesson info modal telling the user to use the generated progress URL to save his exploration progress.",
+	"I18N_SAVE_EXPLORATION_PROGRESS_TEXT_4": "Text displayed on the save progress menu of exploration lesson info modal telling the user that the progress URL generated will save his progress for the next 72 hours only.",
+	"I18N_SAVE_EXPLORATION_PROGRESS_TEXT_5": "Text displayed on the save progress menu of exploration lesson info modal prompting the user to write or copy the genrated URL.",
```
Should these descriptions be gender neutral?

IamEzio replied:
Oh they should be. Fixed them now. Thanks!

nithusha21 reviewed (approved):
1 suggestion below

IamEzio committed:
updated description in qqq.json

gp201 commented on core/templates/pages/exploration-player-page/templates/lesson-information-card-modal.component.spec.ts:
```diff
@@ -253,4 +282,46 @@ describe('Lesson Information card modal component', () => {
+  it('should correctly close save progress menu', () => {
+    componentInstance.saveProgressMenuIsShown = true;
+    componentInstance.closeSaveProgressMenu();
+    expect(componentInstance.saveProgressMenuIsShown).toBeFalse();
```
Please split the test into the following sections (setup, pre-check, action, post-check (as applicable)) by adding line breaks.
Please do this for all the tests.

IamEzio replied:
Done!

gp201 reviewed (changes requested):
one formatting comment. Otherwise similar comments as @kevintab95

IamEzio committed:
added linebreaks in tests

IamEzio commented:
@gp201 PTAL! Thanks

gp201 reviewed (approved):
LGTM

kevintab95 reviewed (approved):
LGTM for codeowner files. Thanks @IamEzio!
